### ISSUE TITLE
Webservice "Add" doesn't seem to work

### ISSUE THREAD

**Issue by techjp** [thirtybees/thirtybees#180]
I have some API code that I have working on PrestaShop 1.6.x sites. This code manages countries, carriers, zones, weights, and related data. Shipping data, essentially.

When I point this code towards a PrestaShop 1.6.x site it operates flawlessly.

When I point the same code at a ThirtyBees 1.0.0 site, UPDATE and DELETE operations work perfectly but ADD operations fail, and I don't seem to be getting any error data. I have tested this trying to create both carriers and zones and the same issue arises.

For zones, the schema info provided by ThirtyBees is identical to that provided by PrestaShop 1.6.x:

```
<thirtybees xmlns:xlink="http://www.w3.org/1999/xlink">
<zone>
<name required="true" maxSize="64" format="isGenericName"/>
<active format="isBool"/>
</zone>
</thirtybees>
```

So I send this XML to my dev site with a request to ADD (properly authenticated of course):

```<?xml version="1.0" encoding="utf-8"?>
<thirtybees>
<zone>
  <name>Zone 1</name>
  <active>1</active>
</zone>
</thirtybees>
```

This should add the new zone to the site and return XML with the newly created zone ID. Instead I get this back:

```<?xml version="1.0" encoding="UTF-8"?>
<thirtybees xmlns:xlink="http://www.w3.org/1999/xlink">
<zones>
</zones>
</thirtybees>
```

Sending the same data in the same way to a PrestaShop site works perfectly. The only difference is of course the PrestaShop site has "prestashop" tags at the start and finish of the XML.

I'm guessing something got changed in the Add routine for the ThirtyBees webservice. Unfortunately I have never managed to figure out which files in the PrestaShop system are used for webservice calls so I'm not sure where to check. 

Am I missing something simple here? Can anyone point me in the right direction for the ThirtyBees webservice files so I can compare them and look for what may have changed?

**Comment by techjp** [comment]
I thought I had tested everything to figure this out but it turns out that some differences in the way that Cloudflare was set up for the PrestaShop store and for the ThirtyBees store is what created this problem. 

I'm not exactly clear on why updates & deletes worked while adds did not but in the end it's not a problem with the API but rather the domain/dns/cdn configuration.

I hope no one spent time trying to troubleshoot this, and if anyone has questions about Cloudflare and the API I'll be happy to try to help.

**Comment by firstred** [comment]
Still using Cloudflare though?

Thanks for telling us what the cause was, I use Cloudflare myself and would
definitely want to check this out.

**Comment by techjp** [comment]
Yes, still using Cloudflare. Love Cloudflare, definitely want to keep using them! I use them for DNS for all my domains and for CDN on active domains. I keep my registrar (namesilo), host (cloudways), dns (cloudflare), and email (google) all on separate services as it makes it *much* easier to change any one service as none are tied together.

The problem I was having is that for my PrestaShop store I was using the naked domain (thisdomain.com) as the main store domain. For the ThirtyBees store I decided to use www.differentdomain.com as the main store domain. The reason for this is that I learned a little more about cookies and using naked domains results in cookies that are "valid" for all subdomains as well. This can be a problem in some situations and is usually not desirable. So despite it not looking as clean, it seems that using www.yourdomain.com is the better choice than just yourdomain.com.

Anyway, I use Cloudflare's Page Rules to enforce certain things. https is enforced, and my desired store domain is also enforced. For the PrestaShop store anyone who visits www.thisdomain.com gets redirected by Cloudflare to https:/\/thisdomain.com. For the ThirtyBees store anyone who visits differentdomain.com gets redirected to https:/\/www.differentdomain.com.

The page rules are what created the problem. I had the API code set up to connect to the naked domain. When it connected cloudflare redirected the connection to the www prefix. This didn't seem to affect update or delete calls but it resulted in add calls failing with no error. It wasn't until I used Fiddler to examine the exact requests that I noticed the redirects happening and put 2 & 2 together.

Once I changed the domain my code was connecting to the problem went away. However it also seems wise to follow [Cloudflare's API use suggestions](https://support.cloudflare.com/hc/en-us/articles/200504045-Using-Cloudflare-with-your-API) for connecting to APIs without risk of getting cached data returned.

**Comment by firstred** [comment]
> However it also seems wise to follow Cloudflare's API use suggestions for connecting to APIs without risk of getting cached data returned.

Which is just a separate page rule for `https://www.differentdomain.com/api*` right?

**Comment by techjp** [comment]
> Which is just a separate page rule for https://www.differentdomain.com/api* right?

Yes. They describe how to set up that rule it in the [blog post](https://support.cloudflare.com/hc/en-us/articles/200504045-Using-Cloudflare-with-your-API) that I also linked to above.

There is some security risk in doing this because it opens that specific URL to more attacks by lowering the security settings and removing it from the Cloudflare cache. While I realize it is "security through obscurity" it might be helpful to give API users the ability to set their own API URL to use with each TB store.

**Comment by firstred** [comment]
> it might be helpful to give API users the ability to set their own API URL to use with each TB store.

Good idea. Currently it works by using a special rewrite rule, but it should become a separate Controller, like any other page, so it routes properly via the Dispatcher class and we no longer need the webservice folder hacks.

**Comment by techjp** [comment]
I saw that rewrite rule in the back office but wasn't sure what would happen if I monkeyed with it. Could users change it there without anything going wrong?

What I'd *really* like to see for the API is bulk updates. It could be done eBay-style by sending giant zipped XML files or just by sending a CSV -- maybe a zipped CSV to make transfers faster. Being able to send 10,000 or 100,000 price updates in a single API call would be hugely beneficial.

This is getting a bit off topic, but I have a few questions that if you have a moment to answer I'd greatly appreciate it:

1. I have never managed to send NULL values through the API. I asked PrestaShop about this in 2015 but never received an answer. Sending NULL through the API (tried a lot of different ways to do this) just results in the API storing a zero in the database, and with some fields that breaks PS until the field is updated with NULLs. Is there a way to send NULL that I haven't figured out?

2. I've seen it mentioned in many places that TB will maintain compatibility with PS 1.6 for TB 1.0.x versions but that compatibility will probably break after that. I have quite a lot of time invested in integrating our in-house software with the PS API and it would suck to have to redo much of that to make use of later versions of TB.  I'm sure I can't be the only person in this position. If necessary changes to the API could be done in such a way as to add features to the existing API schemas without requiring complete rewrites of all client software that would be incredible.

3. I'm sure you're well acquainted with how bad the standard PS search function is, especially for a large site. I've been playing with Elasticsearch and a free module called [Brad](https://github.com/Invertus/brad/). Any chance of integrating this into TB?
